### PR TITLE
Add intrinsics for determining a type's tracing needs

### PR DESCRIPTION
**Pull request by jacob-hughes**
These are useful for libgc, as it can optimise allocation based on these
properties.

### CODE REVIEW

**Comment by jacob-hughes** [comment]
I inverted the logic of `needs_tracing` but forgot to update the test, here's the fix: 2eda2776474adaaca5320e870f243fea36ed7a2d

File: compiler/rustc_middle/src/ty/util.rs [hunk 1]
```diff
@@ -739,6 +739,22 @@ impl<'tcx> ty::TyS<'tcx> {
         tcx_at.is_conservative_raw(param_env.and(self))
     }
 
+    pub fn needs_tracing(
+        &'tcx self,
+        tcx_at: TyCtxtAt<'tcx>,
+        param_env: ty::ParamEnv<'tcx>,
+    ) -> bool {
+        self.is_conservative(tcx_at, param_env) || !self.is_no_trace(tcx_at, param_env)
```
Comment (ltratt) [inline compiler/rustc_middle/src/ty/util.rs:747]: Should `||` be `&&`?

Comment (jacob-hughes) [reply compiler/rustc_middle/src/ty/util.rs:747]: No, `is_conservative` is like an override trait that takes precedence over `NoTrace`.

Comment (ltratt) [reply compiler/rustc_middle/src/ty/util.rs:747]: Dumb question: why would someone need to do such an override?

Comment (jacob-hughes) [reply compiler/rustc_middle/src/ty/util.rs:747]: `NoTrace` implies `alloc_untraceable`.

`!NoTrace` implies `alloc_precise` _or_ `alloc_conservative` but never `alloc_untraceable`.

Until we had `trait Conservative`, there was no single trait that could force the optimiser to always choose the safe-but-slow option (`alloc_conservative`). This is necessary for types like `SOMStack` and also blocks of memory where `layout` doesn't directly correspond to a `T`.

Comment (ltratt) [reply compiler/rustc_middle/src/ty/util.rs:747]: Aha, I think I get it!

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Pull request merged**